kvbc/test/categorization/kv_blockchain_test.cpp

```diff
@@ -1841,6 +1841,188 @@ TEST_F(categorized_kvbc, deletes_ordered_on_add_to_updates) {
   }
 }
 
+// Test root hash calculation and the determinism if the raw block data
+TEST_F(categorized_kvbc, root_hash) {
+  KeyValueBlockchain block_chain{db,
+                                 true,
+                                 std::map<std::string, CATEGORY_TYPE>{{"merkle", CATEGORY_TYPE::block_merkle},
+                                                                      {"merkle2", CATEGORY_TYPE::block_merkle},
+                                                                      {"versioned", CATEGORY_TYPE::versioned_kv},
+                                                                      {"versioned_2", CATEGORY_TYPE::versioned_kv},
+                                                                      {"immutable", CATEGORY_TYPE::immutable},
+                                                                      {"immutable2", CATEGORY_TYPE::immutable},
+                                                                      {"immutable3", CATEGORY_TYPE::immutable}}};
+  // Add block1
+  std::string prev_digest;
+  {
+    Updates updates;
+    BlockMerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key1", "merkle_value1");
+    merkle_updates.addUpdate("merkle_key2", "merkle_value2");
+    updates.add("merkle", std::move(merkle_updates));
+
+    VersionedUpdates ver_updates;
+    ver_updates.calculateRootHash(true);
+    ver_updates.addUpdate("ver_key1", "ver_val1");
+    ver_updates.addUpdate("ver_key2", VersionedUpdates::Value{"ver_val2", true});
+    updates.add("versioned", std::move(ver_updates));
+
+    // Do not calculate root hash
+    VersionedUpdates ver_updates2;
+    ver_updates2.calculateRootHash(false);
+    ver_updates2.addUpdate("ver_key_2_1", "ver_val1");
+    ver_updates2.addUpdate("ver_key_2_2", VersionedUpdates::Value{"ver_val2", true});
+    updates.add("versioned_2", std::move(ver_updates2));
+
+    ASSERT_EQ(block_chain.addBlock(std::move(updates)), (BlockId)1);
+
+    auto rb = block_chain.getRawBlock(1);
+    ASSERT_TRUE(rb.has_value());
+    ASSERT_EQ(rb->data.block_merkle_root_hash.size(), 1);
+    ASSERT_EQ(rb->data.block_merkle_root_hash.count("merkle"), 1);
+    ASSERT_EQ(rb->data.versioned_root_hash.size(), 1);
+    ASSERT_EQ(rb->data.versioned_root_hash.count("versioned"), 1);
+    ASSERT_TRUE(rb->data.immutable_root_hashes.empty());
+    KeyValueBlockchain::KeyValueBlockchain_tester tester;
+    auto cached_rb = tester.getLastRawBlocked(block_chain).second;
+    ASSERT_TRUE(cached_rb.has_value());
+    const auto& ser_rb = categorization::detail::serializeThreadLocal(rb->data);
+    std::string str_rb(ser_rb.begin(), ser_rb.end());
+    const auto& ser_cached_rb = categorization::detail::serializeThreadLocal(*cached_rb);
+    std::string str_cached_rb(ser_cached_rb.begin(), ser_cached_rb.end());
+    ASSERT_EQ(std::hash<std::string>{}(str_rb), std::hash<std::string>{}(str_cached_rb));
```

Comment:
Can we just compare the raw blocks instead of calling std::hash?

Reply:
compare doesn't take into account ordering, I want to detect non-determinism

Reply:
I meant can't we compare the serialized blocks (the std::string objects), something like:
```
ASSERT_EQ(str_rb, str_cached_rb);
```